File: frontend/src/types/binding.ts
export interface PeaBinding {
  id: string
  pea_id: string
  runtime_node_id: string
  driver_instance_id: string
  mappings: TagBinding[]
  validation: BindingValidationSummary
}

export interface TagBinding {
  canonical_tag: string
  driver_tag_id: string
  direction: 'ReadFromDriver' | 'WriteToDriver' | 'Bidirectional'
  transform?: Record<string, unknown> | null
}

export interface BindingValidationSummary {
  valid: boolean
  errors: string[]
  warnings: string[]
}

export interface BindingReadResponse {
  binding_id: string
  canonical_tag: string
  driver_instance_id: string
  driver_tag_id: string
  result: {
    tag_id: string
    tag_name: string
    value: unknown
    error?: number | null
    quality: string
    timestamp: string
  }
}

export interface BindingWriteResponse {
  binding_id: string
  canonical_tag: string
  driver_instance_id: string
  driver_tag_id: string
  actor_id: string
  status: string
  timestamp: string
}

File: backend/shared/src/domain/binding.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PeaBinding {
    pub id: String,
    pub pea_id: String,
    pub runtime_node_id: String,
    pub driver_instance_id: String,
    pub mappings: Vec<TagBinding>,
    pub validation: BindingValidationSummary,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TagBinding {
    pub canonical_tag: String,
    pub driver_tag_id: String,
    pub direction: BindingDirection,
    pub transform: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum BindingDirection {
    ReadFromDriver,
    WriteToDriver,
    Bidirectional,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct BindingValidationSummary {
    pub valid: bool,
    pub errors: Vec<String>,
    pub warnings: Vec<String>,
}
